src/components/TagContainer.ts

```diff
@@ -0,0 +1,277 @@
+            suggestions: getSuggestions.map(suggestion => suggestion.value)
+        });
+    }
+
+    private createTag(newTag: string) {
+        const { afterCreateMicroflow, onChangeMicroflow, mxObject } = this.props;
+        const tagList = this.state.tagList;
+
+        for (const object of this.state.tagCache) {
+            const tagValue = object.get(this.props.tagAttribute) as string;
+            if (newTag === tagValue && newTag.trim() !== "") {
+                tagList.push(newTag);
+                mxObject.addReference(this.referenceAttribute, object.getGuid());
+                this.saveTagData(mxObject);
+                this.setState({ tagList });
+
+                return;
+            }
+        }
+        mx.data.create({
+            callback: object => {
+                object.set(this.props.tagAttribute, newTag);
+                mx.data.commit({
+                    callback: () => {
+                        mxObject.addReference(this.referenceAttribute, object.getGuid());
+                        this.saveTagData(mxObject);
+                        if (afterCreateMicroflow || onChangeMicroflow) {
+                            this.executeAction(mxObject, afterCreateMicroflow);
+                            this.executeAction(mxObject, onChangeMicroflow);
+                        }
+                    },
+                    error: error => window.mx.ui.error("Error occurred attempting to commit: " + error.message),
+                    mxobj: object
+                });
+            },
+            entity: this.tagEntity,
+            error: error => window.mx.ui.error(`Error creating tag object ${this.tagEntity}, ${error.message}`)
+        });
+
+        tagList.push(newTag);
+        this.setState({ tagList });
+    }
+
+    private removeTag(value: string) {
+        if (value.trim() !== "") {
+            const { onChangeMicroflow, mxObject } = this.props;
+            const tagIndex = this.state.tagList.indexOf(value);
+            const tagList = this.state.tagList.slice();
+
+            mx.data.get({
+                callback: (object) => {
+                    if (onChangeMicroflow) {
+                        this.executeAction(mxObject, onChangeMicroflow);
+                    }
+                    mxObject.removeReferences(this.referenceAttribute, object[0].getGuid() as any);
+                    this.saveTagData(mxObject);
+                },
+                error: error => `${error.message}`,
+                xpath: `//${this.tagEntity}[ ${this.props.tagAttribute} = '${value}' ]`
+            });
+
+            tagList.splice(tagIndex, 1);
+            this.setState({
+                alertMessage: "",
+                reloadTagsOnRemove: true,
+                tagList: tagIndex !== -1 ? tagList : this.state.tagList
+            });
+        }
+    }
+
+    private saveTagData(object: mendix.lib.MxObject) {
+        mx.data.commit({
+            mxobj: object,
+            callback: () => undefined
```

Comment:
is this required, because of the typing. or will the Mendix client complain without a callback?